templates/tree-editor/tree/tree-node-factory.ts

```diff
@@ -3,90 +3,87 @@ import { inject, injectable } from 'inversify';
+        }
+        return [];
     }
-    return node;
-  }
 
-  hasCreatableChildren(node: TreeEditor.Node): boolean {
-    return node ? TreeModel.childrenMapping.get(node.jsonforms.type) !== undefined : false;
-  }
+    mapData(data: any, parent?: TreeEditor.Node, property?: string, indexOrKey?: number | string): TreeEditor.Node {
+        if (!data) {
+            this.logger.warn('mapData called without data');
+        }
+
+        const node: TreeEditor.Node = {
+            ...this.defaultNode(),
+            editorId: TreeEditorWidget.EDITOR_ID,
+            name: this.labelProvider.getName(data)!,
+            parent: parent,
+            jsonforms: {
+                type: this.getTypeId(data),
+                data: data,
+                property: property!,
+                index: typeof indexOrKey === 'number' ? indexOrKey.toFixed(0) : indexOrKey
+            }
+        };
+
+        // containments
+        if (parent) {
+            parent.children.push(node);
+            parent.expanded = true;
+        }
+        if (data.children) {
+            const children = data.children as Array<any>;
+            // component types
+            children.forEach((element, idx) => {
+                this.mapData(element, node, 'children', idx);
+            });
+        }
 
-  protected defaultNode(): Pick<
-    TreeEditor.Node,
-    'children' | 'name' | 'jsonforms' | 'id' | 'icon' | 'description' | 'visible' | 'parent' | 'previousSibling' | 'nextSibling' | 'expanded' | 'selected' | 'focus' | 'decorationData'
-  > {
-    return {
-      id: v4(),
-      expanded: false,
-      selected: false,
-      parent: undefined,
-      children: [],
-      decorationData: {},
-      name: '',
-      jsonforms: {
-        type: '',
-        property: '',
-        data: undefined
-      }
+        return node;
     }
-  };
 
-  protected getType(type: string, data: any): string {
-    if (type) {
-        // given typeId
-        return type;
+    hasCreatableChildren(node: TreeEditor.Node): boolean {
+        return node ? CoffeeModel.childrenMapping.get(node.jsonforms.type) !== undefined : false;
     }
-    if (data.typeId) {
-        // typeId of node
-        return data.typeId;
+
+    protected defaultNode(): Pick<
+        TreeEditor.Node,
+        'children' | 'name' | 'jsonforms' | 'id' | 'icon' | 'description' | 'visible' | 'parent' | 'previousSibling' | 'nextSibling' | 'expanded' | 'selected' | 'focus' | 'decorationData'
+    > {
+        return {
+            id: v4(),
+            expanded: false,
+            selected: false,
+            parent: undefined,
+            children: [],
+            decorationData: {},
+            name: '',
+            jsonforms: {
+                type: '',
+                property: '',
+                data: undefined
+            }
+        };
+    }
+
+    /** Derives the type id from the given data. */
+    protected getTypeId(data: any): string {
+        return data.typeId || '';
```

Comment:
In `mapData` a warning is logged if `data` does not exist. However here we are acessing a property which will throw a type error during runtime for this case.
```suggestion
        return data && data.typeId || '';
```